project/webapp/controller/ProductsList.controller.js

```diff
@@ -1,15 +1,193 @@
+
+        oTableBinding.filter(null);
+      },
+
+      _getSearchNameFilter() {
+        const sSearchQuery = this.byId("searchName").getProperty("value");
+
+        return sSearchQuery.length ? new Filter("name", FilterOperator.Contains, sSearchQuery) : null;
+      },
+
+      _getCategoriesFilter() {
+        const aSelectedCategories = this.byId("categorySelect").getProperty("selectedKeys");
+        const aFilters = [];
+
+        if (!!aSelectedCategories.length) {
+          aSelectedCategories.map((sSelectedKey) => {
+            aFilters.push(
+              new Filter({
+                path: "categories",
+                operator: FilterOperator.EQ,
+                value1: sSelectedKey,
+                test: (aCategories) => {
+                  const aResult = aCategories.filter(({id}) => id === sSelectedKey);
+
+                  return !!aResult.length;
+                },
+              })
+            );
+          });
+        }
+
+        return aFilters.length ? aFilters : null;
+      },
+
+      _getDateFilter() {
+        const sDate = this.byId("releaseDate");
+        const sDateStart = sDate.getDateValue();
+        const sDateEnd = sDate.getSecondDateValue();
+
+        if (!sDateStart || !sDateEnd) {
+          return null;
+        }
+
+        return new Filter({
+          path: "releaseDate",
+          operator: FilterOperator.BT,
+          value1: sDateStart.toISOString(),
+          value2: sDateEnd.toISOString(),
         });
+      },
+
+      _getCurrTokens(oEvent) {
+        let aSuppliersTokens = this.byId("supplier").getTokens();
+        const oMultiInput = oEvent.getParameters();
+        const sTokenType = oMultiInput.type;
+
+        if (sTokenType === this.TOKEN_REMOVED_TYPE) {
+          aSuppliersTokens = aSuppliersTokens.filter((item) => item !== oMultiInput.removedTokens[0]);
+        }
+
+        return aSuppliersTokens;
+      },
+
+      _getSupplierFilterWithoutToken() {
+        const sSupplierValue = this.byId("supplier").getValue();
+
+        return [
+          new Filter({
+            path: "suppliers",
+            operator: FilterOperator.EQ,
+            value1: sSupplierValue,
+            test: () => sSupplierValue === "",
+          }),
+        ];
+      },
+
+      _getSupplierFilterWithTokens(aSuppliersTokens) {
+        const aFilters = [];
+
+        aSuppliersTokens.forEach((sToken) => {
+          aFilters.push(
+            new Filter({
+              path: "suppliers",
+              operator: FilterOperator.EQ,
+              value1: sToken,
+              test: (aSuppliers) => {
+                const aResult = aSuppliers.filter(({id}) => id === sToken.getProperty("key"));
+
+                return !!aResult.length;
+              },
+            })
+          );
+        });
+
+        return aFilters;
+      },
+
+      _getSupplierFilter(oEvent) {
+        const aSuppliersTokens = this._getCurrTokens(oEvent);
+
+        if (!aSuppliersTokens.length) {
```

Comment:
it's always easier to read positive conditions than negative, in such cases try to put if (myCondition) {} else {}, instead of  if (!myCondition) {} else {}